Storage/SQL/SQLiteHistory.swift

```diff
@@ -105,6 +148,156 @@ open class SQLiteHistory {
+    if includeBookmarks {
+        ungroupedSQL.append(" AND \(ViewAllBookmarks).url IS NULL")
+    }
+    ungroupedSQL.append(" GROUP BY historyID")
+
+    // Next: limit to only those that have been visited at all within the last six months.
+    // (Don't do that in the innermost: we want to get the full count, even if some visits are older.)
+    // Discard all but the 1000 most frecent.
+    // Compute and return the frecency for all 1000 URLs.
+    let frecenciedSQL =
+        "SELECT *, (\(localFrecencySQL) + \(remoteFrecencySQL)) AS frecency" +
+            " FROM (" + ungroupedSQL + ")" +
+            " WHERE (" +
+            "((localVisitCount > 0) OR (remoteVisitCount > 0)) AND " +                         // Eliminate dead rows from coalescing.
+            "((localVisitDate > \(sixMonthsAgo)) OR (remoteVisitDate > \(sixMonthsAgo)))" +    // Exclude really old items.
+            ") ORDER BY frecency DESC" +
+    " LIMIT 1000"                                 // Don't even look at a huge set. This avoids work.
+
+    // Next: merge by domain and select the URL with the max frecency of a domain, ordering by that sum frecency and reducing to a (typically much lower) limit.
+    // NOTE: When using GROUP BY we need to be explicit about which URL to use when grouping. By using "max(frecency)" the result row
+    //       for that domain will contain the projected URL corresponding to the history item with the max frecency, https://sqlite.org/lang_select.html#resultset
+    //       This is the behavior we want in order to ensure that the most popular URL for a domain is used for the top sites tile.
+    // TODO: make is_bookmarked here accurate by joining against ViewAllBookmarks.
+    // TODO: ensure that the same URL doesn't appear twice in the list, either from duplicate
+    //       bookmarks or from being in both bookmarks and history.
+    let historySQL = [
+        "SELECT historyID, url, title, guid, domain_id, domain,",
+        "max(localVisitDate) AS localVisitDate,",
+        "max(remoteVisitDate) AS remoteVisitDate,",
+        "sum(localVisitCount) AS localVisitCount,",
+        "sum(remoteVisitCount) AS remoteVisitCount,",
+        "max(frecency) AS maxFrecency,",
+        "sum(frecency) AS frecencies,",
+        "0 AS is_bookmarked",
+        "FROM (", frecenciedSQL, ") ",
+        groupClause,
+        "ORDER BY frecencies DESC",
+        "LIMIT \(historyLimit)",
+        ].joined(separator: " ")
+
+    let bookmarksSQL = [
+        "SELECT NULL AS historyID, url, title, guid, NULL AS domain_id, NULL AS domain,",
+        "visitDate AS localVisitDate, 0 AS remoteVisitDate, 0 AS localVisitCount,",
+        "0 AS remoteVisitCount,",
+        "visitDate AS frecencies,",  // Fake this for ordering purposes.
+        "0 as maxFrecency,", // Need this column for UNION
+        "1 AS is_bookmarked",
+        "FROM \(AwesomebarBookmarksTempTable)",
```

Comment:
This table won't exist unless `getOptimizedFrecency` has been called first. So `updateTopSitesCacheQuery` will fail unless the user has started an awesomebar search.

Reply:
Apologies, this and the next comment: my level of sloppiness is really bad near the end of this PR cycle.